api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto

```diff
@@ -0,0 +1,145 @@
+  // or could not be opened, processing continues without error.
+  bool failure_mode_allow = 2;
+
+  // Specifies default options for how HTTP headers, trailers, and bodies are
+  // sent. See ProcessingMode for details.
+  ProcessingMode processing_mode = 3;
+
+  // If true, send each part of the HTTP request or response specified by ProcessingMode
+  // asynchronously -- in other words, send the message on the gRPC stream and then continue
+  // filter processing. If false, which is the default, suspend filter execution after
+  // each message is sent to the remote service and wait up to "message_timeout"
+  // for a reply.
+  bool async_mode = 4;
+
+  // For each property name, select a matching property value to attach to the
+  // request message. Property names are the same ones supported in the WebAssembly
+  // runtime.
+  repeated string request_properties = 5;
+
+  // For each property name, select a matching property value to attach to the
+  // response message.
+  repeated string response_properties = 6;
+
+  // Specifies the timeout for each individual message sent on the stream and
+  // when the filter is running in synchronous mode. Whenever
+  // the proxy sends a message on the stream that requires a response, it will
+  // reset this timer, and will stop processing and return an error (subject
+  // to the processing mode) if the timer expires. There is no timeout when
+  // the filter is running in asynchronous mode. Default is 200 ms.
+  google.protobuf.Duration message_timeout = 7;
+
+  // Specifies if the filter is enabled.
+  //
+  // If :ref:`runtime_key <envoy_api_field_config.core.v3.RuntimeFractionalPercent.runtime_key>` is specified,
+  // Envoy will lookup the runtime key to get the percentage of requests to filter.
+  //
+  // If this field is not specified, the filter will be enabled for all requests.
+  config.core.v3.RuntimeFractionalPercent filter_enabled = 8;
+
+  // Optional additional prefix to use when emitting statistics. This allows to distinguish
+  // emitted statistics between configured *ext_proc* filters in an HTTP filter chain.
+  string stat_prefix = 9;
+}
+
+// Extra settings that may be added to per-route configuration for a
+// virtual host or cluster.
+message ExtProcPerRoute {
+  oneof override {
+    option (validate.required) = true;
+
+    // Disable the filter for this particular vhost or route.
+    // If disabled is specified in multiple per-filter-configs, the most specific one will be used.
+    bool disabled = 1 [(validate.rules).bool = {const: true}];
+
+    // Override aspects of the configuration for this route
+    ExtProcOverrides overrides = 2;
+  }
+}
+
+// Overrides that may be set on a per-route basis
+message ExtProcOverrides {
+  // Set a different processing mode for this route than the default.
+  ProcessingMode processing_mode = 1;
+
+  // Set a different asynchronous processing option than the default.
+  bool async_mode = 2;
```

Comment:
Would it make sense to factor these out with the same settings in the above message?

Reply:
Previously we had an option for the ext_proc server to change the ProcessingMode either permanently or for a certain duration so that the filter could switch between async and sync mode while running. After some discussion we agreed that this extra complexity would be confusing and obviously a race condition. So we changed it so that now ProcessingMode, which specifies which callbacks result in a message to the remote server, is overridable on a per-message basis, but the only way to change async mode is via configuration.

So since ProcessingMode can be overridden for each message, but async mode can only be changed via a configuration change, I think that they need to be separate.